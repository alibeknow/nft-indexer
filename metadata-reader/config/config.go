package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
