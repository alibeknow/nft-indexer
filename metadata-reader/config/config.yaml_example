server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
