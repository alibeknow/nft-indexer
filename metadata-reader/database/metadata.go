package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
