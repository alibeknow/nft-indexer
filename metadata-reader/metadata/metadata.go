package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
