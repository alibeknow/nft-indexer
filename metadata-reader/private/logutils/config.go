package logutils

type Config struct {
	Level string
}
