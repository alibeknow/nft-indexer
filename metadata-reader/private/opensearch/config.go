package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
