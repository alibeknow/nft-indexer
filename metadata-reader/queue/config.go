package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
