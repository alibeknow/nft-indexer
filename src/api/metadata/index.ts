export * from './metadata.controller';
export * from './metadata.module';
export * from './metadata.service';
