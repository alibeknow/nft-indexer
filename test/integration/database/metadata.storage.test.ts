import { v4 as uuidv4 } from 'uuid';
import { ConfigService } from '@nestjs/config';
import { MetadataRepoStorage, UriProtocol } from '@shared/metadata';
import { IBaseConfig } from '@shared/baseconfig';
import { S3Provider } from '@shared/aws';

describe('Metadata Storage repo', () => {
  let s3Provider: S3Provider<IBaseConfig>;
  let bucketName: string;
  beforeAll(async () => {
    s3Provider = new S3Provider(new ConfigService<IBaseConfig, false>());
    bucketName = `metadata-${uuidv4()}`;
    await s3Provider.storage.createBucket({ Bucket: bucketName }).promise();
  });

  describe('[save] method', () => {
    it('should save a metadata document to MongoDB in metadata collection', async () => {
      const metadataRepo = new MetadataRepoStorage(s3Provider, bucketName);
      const metadataItems = [
        {
          id: 'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
          metadata: JSON.stringify({
            description: 'Born out of a bad batch of fermented ...',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
          metadata: JSON.stringify({
            name: 'Braindom #2',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
          metadata: JSON.stringify({
            name: 'NIKE  APE #114',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg',
          }),
        },
      ];

      await Promise.all(metadataItems.map(({ id, metadata }) => metadataRepo.save(id, metadata, UriProtocol.HTTPS)));

      let objectExist;
      objectExist = await s3Provider.objectExist(bucketName, 'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e.json');
      expect(objectExist).toBe(true);
      objectExist = await s3Provider.objectExist(bucketName, 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b.json');
      expect(objectExist).toBe(true);
      objectExist = await s3Provider.objectExist(bucketName, 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72.json');
      expect(objectExist).toBe(true);
    });
  });

  describe('[getWhereIdIn] method', () => {
    it('should retrieve metadata documents by ids', async () => {
      const metadataRepo = new MetadataRepoStorage(s3Provider, bucketName);
      const metadataItems = [
        {
          id: 'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
          metadata: JSON.stringify({
            description: 'Born out of a bad batch of fermented ...',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
          type: UriProtocol.IPFS,
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
          metadata: JSON.stringify({
            name: 'Braindom #2',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
          type: UriProtocol.HTTPS,
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
          metadata: JSON.stringify({
            name: 'NIKE  APE #114',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg',
          }),
          type: UriProtocol.HTTPS,
        },
      ];

      await Promise.all(metadataItems.map(({ id, metadata, type }) => metadataRepo.save(id, metadata, type)));

      const results1 = await metadataRepo.getWhereIdIn([
        'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
      ]);

      const results2 = await metadataRepo.getWhereIdIn([
        'eth:fake-contract-id:fake-token-id',
      ]);

      const results3 = await metadataRepo.getWhereIdIn([
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
        'eth:fake-contract-id:fake-token-id',
      ]);

      const results4 = await metadataRepo.getWhereIdIn([
        'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
      ]);

      expect(results1).toHaveLength(2);
      expect(results1).toMatchSnapshot();
      expect(results2).toHaveLength(0);
      expect(results2).toMatchSnapshot();
      expect(results3).toHaveLength(1);
      expect(results3).toMatchSnapshot();
      expect(results4).toHaveLength(3);
      expect(results4).toMatchSnapshot();
    });
  });
});
