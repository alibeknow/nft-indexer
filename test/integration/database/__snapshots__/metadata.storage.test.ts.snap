// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata Storage repo [getWhereIdIn] method should retrieve metadata documents by ids 1`] = `
[
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e",
    "metadata": "{"description":"Born out of a bad batch of fermented ...","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "ipfs",
  },
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72",
    "metadata": "{"name":"NIKE  APE #114","description":"COUNTDOWN OVER. MINTING LIVE.","image":"http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg"}",
    "type": "https",
  },
]
`;

exports[`Metadata Storage repo [getWhereIdIn] method should retrieve metadata documents by ids 2`] = `[]`;

exports[`Metadata Storage repo [getWhereIdIn] method should retrieve metadata documents by ids 3`] = `
[
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b",
    "metadata": "{"name":"Braindom #2","description":"COUNTDOWN OVER. MINTING LIVE.","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "https",
  },
]
`;

exports[`Metadata Storage repo [getWhereIdIn] method should retrieve metadata documents by ids 4`] = `
[
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e",
    "metadata": "{"description":"Born out of a bad batch of fermented ...","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "ipfs",
  },
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b",
    "metadata": "{"name":"Braindom #2","description":"COUNTDOWN OVER. MINTING LIVE.","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "https",
  },
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72",
    "metadata": "{"name":"NIKE  APE #114","description":"COUNTDOWN OVER. MINTING LIVE.","image":"http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg"}",
    "type": "https",
  },
]
`;
